internal/client/eth_client.go

```diff
@@ -23,31 +30,84 @@ type EthClientInterface interface {
 	BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error)
 }
 
-// NewEthClient ... Initializer
-func NewEthClient() EthClientInterface {
+func NewEthClient(restyClient *resty.Client) EthClientInterface {
+	if restyClient == nil {
+		restyClient = resty.New()
+	}
+
+	restyDebug, _ := strconv.ParseBool(os.Getenv("RESTY_DEBUG"))
+	restyClient.SetDebug(restyDebug)
+
 	return &EthClient{
-		client: &ethclient.Client{},
+		client:      &ethclient.Client{},
+		restyClient: restyClient,
 	}
 }
 
-// DialContext ... Wraps go-etheruem node dialContext RPC creation
 func (ec *EthClient) DialContext(ctx context.Context, rawURL string) error {
 	client, err := ethclient.DialContext(ctx, rawURL)
 
 	if err != nil {
 		return err
 	}
 
+	// Configure retry logic
+	// to-do make env variables
+	retryCount := 3
+	retryWaitTime := 5
+	retryMaxWaitTime := 20
+
+	ec.restyClient.
+		SetRetryCount(retryCount).
+		SetRetryWaitTime(time.Duration(retryWaitTime) * time.Second).
+		SetRetryMaxWaitTime(time.Duration(retryMaxWaitTime) * time.Second).
+		AddRetryCondition(
+			func(r *resty.Response, err error) bool {
+				return r.IsError() || err != nil
+			},
+		)
+
 	ec.client = client
+	ec.url = rawURL // Store the URL
 	return nil
 }
 
-// HeaderByNumber ... Wraps go-ethereum node headerByNumber RPC call
 func (ec *EthClient) HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error) {
-	return ec.client.HeaderByNumber(ctx, number)
+	var header *types.Header
+	resp, err := ec.restyClient.R().
+		SetContext(ctx).
+		SetPathParams(map[string]string{
+			"blockNumber": number.String(),
+		}).
+		Get(ec.url + "/eth_getBlockByNumber/{blockNumber}")
+
```

Comment:
I don't think we should be deprecating go-ethereum client use in favor of a retryable client. Is there no way to bind a retry-able client into the [ethclient](https://pkg.go.dev/github.com/ethereum/go-ethereum/ethclient) directly?

Reply:
I can use the retry-go package to add a 'try again' feature. It would basically ask for an Ethereum block header. If it fails due to some error, instead of giving up, it tries again a few more times.
Would look like this
`func (ec *EthClient) HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error) {
	retryableFunc := func() (*types.Header, error) {
		header, err := ec.client.HeaderByNumber(ctx, number)
		if err != nil {
			return nil, err
		}
		return header, nil
	}

	// Execute the function with retry logic
	var header *types.Header
	err := retry.Do(
		func() error {
			var err error
			header, err = retryableFunc()
			return err
		},
		retry.Attempts(3),
		retry.Delay(5*time.Second),
		retry.MaxDelay(20*time.Second),
	)

	if err != nil {
		return nil, err
	}
	return header, nil
}`